base/src/main/java/org/mozilla/jss/ssl/SSLSocket.java

```diff
@@ -761,7 +761,7 @@ public synchronized int getSoTimeout() throws SocketException {
      * Closes this socket.
      */
     @Override
-    public void close() throws IOException {
+    public synchronized void close() throws IOException {
         synchronized (this) {
```

Comment:
I am sceptical of the thread safety paradigm used in this class, and I think there is deadlock potential.

We have a `readLock` object and a `writeLock` object, as well as obtaining a lock on the instance itself and the `close` method code itself is locked. In the `read` method, we obtain the lock on `readLock`, then we immediately take the lock on the instance, so what is the point of `readLock`? If we have the instance lock then no reading or writing can happen, so the granularity is unnecessary?

The `close` method takes the instance lock, then does "stuff", then potentially takes the `readLock` then the `writeLock`. The `write` and `read` methods do it the other way around, they take the `read/writeLock`, THEN they try to take the instance lock. If a `read` call came in just after a `close` call, then `close` holds the instance lock, waiting for `readLock`, and `read` has `readLock`, waiting for the instance lock. Deadlock.

What do you think @edewata?

Reply:
@ckelleyRH I have review again all the changes and I agree in this case to synchronize the method without change the logic could create a deadlock problem. I would remove the `synchronized` from the method and add the exception in sonarcloud for the moment. I'll keep only for the others if you agree.

Reply:
I agree with the possibility of deadlock here. In general I'd prefer we use only `synchronized` methods instead of `synchronized(object)` because the former is easier to understand. Maybe we should convert all `synchronized(this)` into new `synchronized` methods?

For the other classes, does it mean we need to synchronize all `hashCode()` and `equals()`? I think in general we assume that the classes are not thread-safe by default, so usually we only add `synchronized` if it's necessary.